### PR TITLE
UIP in SProp

### PR DESCRIPTION
**Pull request by SkySkimmer**
TODO:
- fix alternate reductions (?)

Overlays: https://github.com/unicoq/unicoq/pull/31 https://github.com/Mtac2/Mtac2/pull/216 https://github.com/LPCIC/coq-elpi/pull/68 https://github.com/mattam82/Coq-Equations/pull/224 https://github.com/coq-community/paramcoq/pull/36 https://github.com/damien-pous/relation-algebra/pull/9 https://github.com/Karmaki/coq-dpdgraph/pull/66 https://github.com/lukaszcz/coqhammer/pull/39 https://github.com/MetaCoq/metacoq/pull/434

### CODE REVIEW

**Comment by JasonGross** [comment]
It seems strange to me that UIP of sprop equality gives UIP of Prop equality.  What causes this, and that does the HoTT flag disable?

**Comment by SkySkimmer** [comment]
>It seems strange to me that UIP of sprop equality gives UIP of Prop equality. What causes this, and that does the HoTT flag disable?

You can define `f : eq a b -> seq a b` and `g : seq a b -> eq a b` with `g (f p) = p`, then `f p = f q` by sprop so `p = g (f p) = g (f q) = q`

On second thought -indices-matter does what I want so there may be no need for a new flag, unless someone wants to have identities in Prop without UIP.

**Comment by andres-erbsen-sifive** [comment]
as of before this change, was it not possible to prove that any two proofs of the same `seq` are equal? I thought the entire point of SProp was that proofs are convertible.

**Comment by SkySkimmer** [comment]
>as of before this change, was it not possible to prove that any two proofs of the same seq are equal? I thought the entire point of SProp was that proofs are convertible.

It was possible, however seq was a weaker statement than eq.

**Comment by SkySkimmer** [comment]
Bench is not great
~~~

┌────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬───────────────────────────────┬───────────────────────────┐
│                        │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]     │        mem faults         │
│                        │                         │                                             │                                             │                               │                           │
│           package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD PDIFF   │     NEW     OLD   PDIFF   │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│       coq-math-classes │  217.95  217.78 +0.08 % │      608237830980      608025709972 +0.03 % │      767963442228      767142083146 +0.11 % │     513656     513664 -0.00 % │      44      20 +120.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│ coq-mathcomp-ssreflect │   40.05   39.98 +0.18 % │      110704560148      110343847162 +0.33 % │      133881431626      133185436518 +0.52 % │     521460     520372 +0.21 % │      12      61  -80.33 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│               coq-hott │  336.16  335.25 +0.27 % │      915707484558      911289006389 +0.48 % │     1386507853816     1383250690928 +0.24 % │     476528     476552 -0.01 % │     472     246  +91.87 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│               coq-corn │ 1460.44 1455.51 +0.34 % │     4077947090988     4063989423131 +0.34 % │     6025817695426     5999824371955 +0.43 % │     881296     892612 -1.27 % │       1      77  -98.70 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│  coq-mathcomp-fingroup │   51.52   51.31 +0.41 % │      143638102088      142987158250 +0.46 % │      186286998785      185151246638 +0.61 % │     562380     562364 +0.00 % │      91       7 +1200.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│            coq-unimath │ 3004.22 2987.81 +0.55 % │     8371779234380     8324420961898 +0.57 % │    14978391128021    14926910297261 +0.34 % │    3661112    3548824 +3.16 % │     596     466  +27.90 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│   coq-mathcomp-algebra │  165.24  164.28 +0.58 % │      460744687116      458405544290 +0.51 % │      568436352054      565186038301 +0.58 % │     610324     609732 +0.10 % │      16       0    +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│              coq-color │  700.91  696.49 +0.63 % │     1960553516237     1948374036995 +0.63 % │     2332796375755     2321810325889 +0.47 % │    1381808    1382560 -0.05 % │     235     234   +0.43 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│          coq-fourcolor │ 2179.05 2165.26 +0.64 % │     6074940722740     6039260202322 +0.59 % │    11370023210293    11351690128073 +0.16 % │     927424     927172 +0.03 % │       1       0    +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│          coq-fiat-core │  108.54  107.84 +0.65 % │      309380064126      308453045153 +0.30 % │      385840033002      384481201688 +0.35 % │     484716     484572 +0.03 % │     393     233  +68.67 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│       coq-fiat-parsers │  639.74  635.44 +0.68 % │     1789913199684     1777413182533 +0.70 % │     2699654563489     2683286163751 +0.61 % │    3211944    3231800 -0.61 % │     528     663  -20.36 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│             coq-sf-plf │   44.10   43.74 +0.82 % │      123468527172      122539141154 +0.76 % │      159273381757      158340314305 +0.59 % │     490660     481116 +1.98 % │      89      19 +368.42 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│ coq-mathcomp-character │  182.37  180.78 +0.88 % │      507721592740      503652180746 +0.81 % │      668129658478      662605779276 +0.83 % │     849088     842752 +0.75 % │       0      11 -100.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│         coq-coquelicot │   75.99   75.29 +0.93 % │      208921394288      207074100990 +0.89 % │      252618433700      250099281427 +1.01 % │     655308     653068 +0.34 % │     444     142 +212.68 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│  coq-mathcomp-solvable │  190.27  188.23 +1.08 % │      530873726286      525243270595 +1.07 % │      708748447234      700087577414 +1.24 % │     784452     786340 -0.24 % │       9      17  -47.06 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│            coq-bignums │   69.82   69.04 +1.13 % │      193949060348      192222374280 +0.90 % │      255073196451      252778170117 +0.91 % │     492912     491140 +0.36 % │     178     145  +22.76 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│         coq-verdi-raft │ 1334.70 1319.68 +1.14 % │     3723259698651     3681443829178 +1.14 % │     5034031140340     4986774689057 +0.95 % │    1812392    1811192 +0.07 % │       0       0    +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│              coq-flocq │  463.84  458.10 +1.25 % │     1291143096357     1274946705314 +1.27 % │     1780229475994     1773408957910 +0.38 % │    1163620    1161464 +0.19 % │     281      74 +279.73 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│     coq-mathcomp-field │  232.55  229.56 +1.30 % │      648248379216      640405481048 +1.22 % │      923173973096      913958664695 +1.01 % │     745700     746612 -0.12 % │       0      75 -100.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│        coq-fiat-crypto │ 3836.86 3782.08 +1.45 % │    10657622802812    10500105910583 +1.50 % │    16664231748892    16476051692552 +1.14 % │    2334020    2341364 -0.31 % │    1169     733  +59.48 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│        coq-lambda-rust │ 1560.58 1534.83 +1.68 % │     4351532492092     4277301311848 +1.74 % │     5750097885420     5658005580098 +1.63 % │    1488764    1526244 -2.46 % │       0       0    +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│             coq-geocoq │ 1970.90 1936.77 +1.76 % │     5499823659928     5406826395001 +1.72 % │     7913760965635     7827202355417 +1.11 % │    1253824    1253924 -0.01 % │     148     214  -30.84 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│ coq-mathcomp-odd-order │ 1469.70 1443.68 +1.80 % │     4097067845402     4023341181720 +1.83 % │     6945756642947     6865934126922 +1.16 % │    1263636    1252560 +0.88 % │     158      46 +243.48 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│              coq-verdi │  121.00  118.18 +2.39 % │      336327175254      329305412513 +2.13 % │      426214954771      417107005360 +2.18 % │     557236     559536 -0.41 % │       8       6  +33.33 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────────────┤
│ coq-fiat-crypto-legacy │ 6449.70 6270.40 +2.86 % │    17959570538672    17454770201690 +2.89 % │    29120223661135    28403347658290 +2.52 % │    2418956    2339932 +3.38 % │    1011    1243  -18.66 % │
└────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴───────────────────────────────┴───────────────────────────┘
~~~

**Comment by ejgallego** [comment]
@SkySkimmer how about flambda ?

**Comment by SkySkimmer** [comment]
>@SkySkimmer how about flambda ?

Can you start the bench, I don't know how to get flambda in there.

**Comment by ejgallego** [comment]
> Can you start the bench, I don't know how to get flambda in there.

Started https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/737

**Comment by SkySkimmer** [comment]
>Started https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/737

Didn't work

**Comment by ejgallego** [comment]
Ups I was careless and didn't update the ref to your tree, please relaunch.

**Comment by ejgallego** [comment]
https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/738/ should now have the right hash.

**Comment by mattam82** [comment]
Can you give a quick summary of the changes and the meaning of CaseInvert?

**Comment by SkySkimmer** [comment]
We allow inductives type with 1 constructor which has no arguments in SProp without squashing. Since equality fits that condition we get UIP.

Problem: when you have UIP, you find that if `x == y` then `@eq_rect A x P v y e = @eq_rect A x P v y (eq_refl x) = v` regardless of `e` (because by irrelevance `e == eq_refl x`, and because `x == y` this is well typed).
We need this to be in the reduction, as for instance we could have `f : @eq_rect A x (fun _ => Type) (nat -> nat) y e` which means `f` has a type which converts to an arrow and `f 0` is well typed, exactly when `x == y`.
CaseInvert is there to implement this special kind of conversion dependent reduction rule (for matches instead of the eliminators I used in the example ofc). The rule is that for inductive types in SProp with 1 constructor, when matching to produce a relevant result instead of producing the regular Case we produce CaseInvert, and CaseInvert reduces the its only branch when the indices of the scrutinee are convertible to those of the constructor instantiated by the appropriate universes and parameters. (for matches from irrelevant to irrelevant we don't need to do special reduction, so we don't)
CaseInvert carries extra data when compared to Case (the `case_invert` argument) to make it possible to get the indices of the scrutinee without retyping. While we're at it we also put in the universes and parameters so that we don't have to hack them out of the predicate/branch.
Some APIs like inductiveops are changed a bit to make it easier to generate the appropriate `case_invert`.

**Comment by mattam82** [comment]
Ok, that makes sense. What about the vm/native compute? How do you explain the time difference for things that don’t use it?

**Comment by mattam82** [comment]
Also, True : SProp also fits the criterion, right, and any phantom type ˋInd foo : ty -> SProp := isFoo : foo (tyconstructor)` as well?

**Comment by SkySkimmer** [comment]
yes

Gaëtan Gilbert

**Comment by SkySkimmer** [comment]
Looks like having equality in SProp and impredicative Prop breaks normalization.

Example from Thierry Coquand:
~~~coq
Set Allow StrictProp.

Inductive seq {A} a : A -> SProp := srefl : seq a a.
Arguments srefl {_ _}, {_} _.

Axiom all_eq : forall P Q : Prop, seq P Q.

Definition K (A B : Prop) (x:A): B := match all_eq A B with srefl => x end.

Definition bot : Prop := forall P : Prop, P.

Definition c : bot := fun X : Prop => K (bot -> bot) X (fun z : bot => z (bot -> bot) z).

Eval lazy in c (bot -> bot) c.
(* loops *)
~~~
(using SProp instead of Prop terminates because case inversion is only done for relevant matches, I guess with -impredicative-set you can use Set instead of Prop)

**Comment by maximedenes** [comment]
Wow, this is a very nice example. The critical idea seems to be that after instantiation, only a trivial cast is used (`bot -> bot`) to (`bot -> bot`), so the match (in `K`) can be inverted, if I understood the reduction rules correctly.

**Comment by ppedrot** [comment]
According to Thierry, the example originally comes from Girard's paper on System F and is fundamentally tied to the use of impredicativity.

**Comment by maximedenes** [comment]
> is fundamentally tied to the use of impredicativity

Yes, this seems to be the case. In fact, you can see it as `Definition unit := forall P : Prop, P -> P` being very close to `unit := unit -> unit`, modulo instantiation. In the usual CIC, casting from a concrete type to an abstract one will require a stuck match leading to a commutative cut that will not reduce. The reduction rules introduced here makes these matches disappear.

Without impredicativity, you'd need to cast between two actually different types, so the matches won't be reduced even with the new rules (since the indices have to agree).

**Comment by SkySkimmer** [comment]
OK it loops even without using Prop, we just need to box up a bit in the transport lemma:
~~~coq
Set Allow StrictProp.

Record Box (A:SProp) : Set := box { unbox : A }.
Arguments box {_} _. Arguments unbox {_} _.

Inductive seq {A} a : A -> SProp := srefl : seq a a.
Arguments srefl {_ _}, {_} _.

Axiom all_eq : forall P Q : SProp, seq P Q.

Definition K (A B : SProp) (x:A): B := unbox match all_eq A B with srefl => box x end.

Definition bot : SProp := forall P : SProp, P.

Definition c : bot := fun X : SProp => K (bot -> bot) X (fun z : bot => z (bot -> bot) z).

Eval lazy in c (bot -> bot) c.
~~~

**Comment by mattam82** [comment]
I guess I always thought about strict equalities of terms rather than types. Could the inconsistency also arise if you restrict seq to A : Set (and Set not impredicative)?

**Comment by SkySkimmer** [comment]
Version with a consistent axiom
~~~coq
Axiom all_eq : forall (P Q:Prop), P -> Q -> seq P Q.

Definition transport (P Q:Prop) (x:P) (y:Q) : Q
  := match all_eq P Q x y with srefl => x end.

Definition top : Prop := forall P : Prop, P -> P.

Definition c : top :=
  fun P p => transport (top -> top) P
                    (fun x : top =>
                       x (top -> top) (fun x => x) x) p.

Eval lazy in c (top -> top) (fun x => x) c.
~~~

**Comment by JasonGross** [comment]
Can this infinite loop be exploited to produce a proof of False (or a proof that this axiom is inconsistent)?

**Comment by SkySkimmer** [comment]
We have consistency through ett

**Comment by ppedrot** [comment]
@JasonGross essentially we lose decidability of type-checking, but consistency is not endangered. Actually it even seems that we still have canonicity⁽¹⁾ in this system.

⁽¹⁾ For a good notion of canonicity.

**Comment by ppedrot** [comment]
Shouldn't we close this for the time being?

**Comment by SkySkimmer** [comment]
Coq call: guard by a flag (warning users that it breaks termination) and merge.

**Comment by SkySkimmer** [comment]
I need to rebase all the overlays but other than that it should be ready.

File: kernel/constr.ml [hunk 1]
```diff
@@ -99,7 +103,7 @@ type ('constr, 'types, 'sort, 'univs) kind_of_term =
   | Const     of (Constant.t * 'univs)
   | Ind       of (inductive * 'univs)
   | Construct of (constructor * 'univs)
-  | Case      of case_info * 'constr * 'constr * 'constr array
+  | Case      of case_info * 'constr * ('constr, 'univs) case_invert * 'constr * 'constr array
```
Comment (ejgallego) [inline kernel/constr.ml]: Just wondering if this would have been a good opportunity to use an inline record.

Comment (SkySkimmer) [reply kernel/constr.ml]: I felt like it would be too painful compared to adding a `_, ` in most places.

**Comment by SkySkimmer** [comment]
I broke the doc somehow, maybe @coq/doc-maintainers can figure it out.

**Review by Zimmi48 — APPROVED** [review]
The documentation looks very good to me.

The Sphinx error is undecipherable, but given that the exception is raised just after `parsing bibtex file /builds/coq/coq/doc/sphinx/biblio.bib... parsed 55 entries`, I think it is likely that it has something to do with the format of the bibtex entry. Maybe the `sphinx-bibtex` extension does not support the `@online` entry type. I see that previous HAL pre-prints have been inserted as `@techreport`.

**Comment by SkySkimmer** [comment]
All failures are spurious.

**Comment by SkySkimmer** [comment]
@mattam82 asked for a ping offline

**Comment by Zimmi48** [comment]
@SkySkimmer @maximedenes What is the status of this? Please retarget the milestone as appropriate.

**Review by maximedenes — APPROVED** [review]
LGTM, I think we can address alternate reductions in a next step, if there's user demand for it.

**Comment by maximedenes** [comment]
This PR has been waiting for a year now, and discussed during a Coq call. I'll merge soon if no objections (probably after primitive arrays, to minimize conflicts).

File: pretyping/reductionops.ml [hunk 2]
```diff
@@ -536,12 +537,13 @@ let reduce_and_refold_cofix recfun env sigma cofix sk =
...
     | Construct ((ind_sp,i),u) ->
-(*	let ncargs = (fst mia.mci).(i-1) in*)
+(*      let ncargs = (fst mia.mci).(i-1) in*)
         let real_cargs = List.skipn mia.mci.ci_npar mia.mcargs in
         applist (mia.mlf.(i-1),real_cargs)
     | CoFix cofix ->
         let cofix_def = contract_cofix sigma cofix in
-        mkCase (mia.mci, mia.mP, applist(cofix_def,mia.mcargs), mia.mlf)
+        (* XXX Is NoInvert OK here? *)
+        mkCase (mia.mci, mia.mP, NoInvert, applist(cofix_def,mia.mcargs), mia.mlf)
```
Comment (SkySkimmer) [inline pretyping/reductionops.ml:546]: This case of a comatch with special reduction can probably happen, not sure if using NoInvert is a problem (maybe the result of this function gets immediately reduced? I don't understand this code)
Maybe we should just ban coinductives in sprop.

File: proofs/logic.ml [hunk 3]
```diff
@@ -397,14 +398,15 @@ let rec mk_refgoals ~check env sigma goalacc conclty trm =
         let ty = EConstr.Unsafe.to_constr ty in
           (acc',ty,sigma,c)
 
-      | Case (ci,p,c,lf) ->
+      | Case (ci,p,iv,c,lf) ->
+        (* XXX Is ignoring iv OK? *)
```
Comment (SkySkimmer) [inline proofs/logic.ml:402]: I have no clue what this function and the one a few lines below do, so I don't know if the case_invert data should be ignored or something should be done with it.

**Review by SkySkimmer — COMMENTED** [review]
Handled most remaining todos. Remain:
- VM/native ignore the case_invert in compilation, if the match gets stuck readback in pretyping will user_err (proper readback is probably feasible, full reduction is way harder).
- see other github comments attached to code

File: kernel/constr.mli [hunk 4]
```diff
@@ -558,7 +575,7 @@ val compare_head : constr constr_compare_fn -> constr constr_compare_fn
 (** Convert a global reference applied to 2 instances. The int says
    how many arguments are given (as we can only use cumulativity for
    fully applied inductives/constructors) .*)
-type 'univs instance_compare_fn = GlobRef.t -> int ->
+type 'univs instance_compare_fn = (GlobRef.t * int) option ->
```
Comment (SkySkimmer) [inline kernel/constr.mli:578]: BTW this change is the same in persistent arrays: we need to be able to compare universes attached to no globref.

**Comment by SkySkimmer** [comment]
Overlays rebased (mtac2 is missing mtac2/mtac2#272)

**Comment by XVilka** [comment]
Only one metacoq failure:
```
File "src/g_metacoq_safechecker.mlg", line 73, characters 6-261:
Warning 8: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
UIP _
File "src/g_metacoq_safechecker.ml", line 1:
Error: Some fatal warnings were triggered (1 occurrences)
Command exited with non-zero status 2
src/g_metacoq_safechecker.cmx (real: 0.42, user: 0.08, sys: 0.02, mem: 34968 ko)
Makefile.plugin:667: recipe for target 'src/g_metacoq_safechecker.cmx' failed
```

**Comment by SkySkimmer** [comment]
All green!

**Pull request merged**